File: orchestrator/transforms.py
"""Compile transform JSON (from mapping edges) into DuckDB SQL expressions.

Transforms make the two sides comparable; they are NOT judgments. A transform is
either a single ``{"op": ...}`` dict or a list of them applied left-to-right.

Supported ops mirror docs/architecture/rule-engine-spec.md §3.
"""
from __future__ import annotations
from typing import Optional, Union, List, Dict, Any

from .quoting import lit, num

Transform = Optional[Union[Dict[str, Any], List[Dict[str, Any]]]]


def compile_transform(expr: str, transform: Transform) -> str:
    """Wrap ``expr`` (already a valid SQL expression, e.g. a quoted column) with the
    transform chain and return the resulting SQL expression."""
    if not transform:
        return expr
    ops = transform if isinstance(transform, list) else [transform]
    for op in ops:
        expr = _apply(expr, op)
    return expr


def _apply(expr: str, op: Dict[str, Any]) -> str:
    name = op.get("op")
    args = op.get("args", [])
    if name == "trim":
        return f"trim({expr})"
    if name == "upper":
        return f"upper({expr})"
    if name == "lower":
        return f"lower({expr})"
    if name == "round":
        return f"round(TRY_CAST({expr} AS DOUBLE), {int(args[0])})"
    if name == "scale":
        return f"(TRY_CAST({expr} AS DOUBLE) / {num(args[0])})"
    if name == "substr":
        return f"substr({expr}, {int(args[0])}, {int(args[1])})"
    if name == "pad_left":
        return f"lpad({expr}, {int(args[0])}, {lit(args[1])})"
    if name == "replace":
        return f"replace({expr}, {lit(op['from'])}, {lit(op['to'])})"
    if name == "null_if":
        return f"nullif({expr}, {lit(args[0])})"
    if name == "date_format":
        return f"strftime(try_strptime({expr}, {lit(op['from'])}), {lit(op['to'])})"
    raise ValueError(f"unknown transform op: {name!r}")

File: dotnet/ValidationCenter.RuleEngine/Transforms.cs
using System.Globalization;
using System.Text.Json;

namespace ValidationCenter.RuleEngine;

/// <summary>Compile transform JSON into DuckDB SQL expressions. Mirrors
/// orchestrator/transforms.py.</summary>
public static class Transforms
{
    public static string Compile(string expr, JsonElement? transform)
    {
        if (transform is null) return expr;
        var t = transform.Value;
        if (t.ValueKind == JsonValueKind.Null) return expr;
        if (t.ValueKind == JsonValueKind.Array)
        {
            foreach (var op in t.EnumerateArray()) expr = Apply(expr, op);
            return expr;
        }
        return Apply(expr, t);
    }

    private static string S(JsonElement op, string name) => op.GetProperty(name).GetString()!;
    private static JsonElement Arg(JsonElement op, int i) => op.GetProperty("args")[i];
    private static int ArgI(JsonElement op, int i) => Arg(op, i).GetInt32();
    private static string ArgS(JsonElement op, int i) => Arg(op, i).GetString()!;
    private static string ArgN(JsonElement op, int i) =>
        Arg(op, i).GetDouble().ToString("R", CultureInfo.InvariantCulture);

    private static string Apply(string expr, JsonElement op)
    {
        var name = op.GetProperty("op").GetString();
        return name switch
        {
            "trim" => $"trim({expr})",
            "upper" => $"upper({expr})",
            "lower" => $"lower({expr})",
            "round" => $"round(TRY_CAST({expr} AS DOUBLE), {ArgI(op, 0)})",
            "scale" => $"(TRY_CAST({expr} AS DOUBLE) / {ArgN(op, 0)})",
            "substr" => $"substr({expr}, {ArgI(op, 0)}, {ArgI(op, 1)})",
            "pad_left" => $"lpad({expr}, {ArgI(op, 0)}, {Quoting.Lit(ArgS(op, 1))})",
            "replace" => $"replace({expr}, {Quoting.Lit(S(op, "from"))}, {Quoting.Lit(S(op, "to"))})",
            "null_if" => $"nullif({expr}, {Quoting.Lit(ArgS(op, 0))})",
            "date_format" =>
                $"strftime(try_strptime({expr}, {Quoting.Lit(S(op, "from"))}), {Quoting.Lit(S(op, "to"))})",
            _ => throw new ArgumentException($"unknown transform op: {name}")
        };
    }
}
